cache/azblob/azblob.go

```diff
@@ -0,0 +1,299 @@
+	// basepath
+	basePath := ""
+	basePath, err = config.String(ConfigKeyBasepath, &basePath)
+	if err != nil {
+		return nil, err
+	}
+
+	azCache.Basepath = basePath
+
+	readOnly := false
+	readOnly, err = config.Bool(ConfigKeyReadOnly, &readOnly)
+	if err != nil {
+		return nil, err
+	}
+
+	azCache.ReadOnly = readOnly
+
+	// credentials
+	acctName := ""
+	acctName, err = config.String(ConfigKeyAzureAccountName, &acctName)
+	if err != nil {
+		return nil, err
+	}
+
+	acctKey := ""
+	acctKey, err = config.String(ConfigKeyAzureSharedKey, &acctKey)
+	if err != nil {
+
+	}
+
+	var cred azblob.Credential
+
+	if acctName+acctKey == "" {
+		cred = azblob.NewAnonymousCredential()
+		azCache.ReadOnly = true
+	} else {
+		if acctName == "" || acctKey == "" {
+			return nil, fmt.Errorf("both %s and %s must be present", ConfigKeyAzureAccountName, ConfigKeyAzureSharedKey)
+		}
+
+		cred = azblob.NewSharedKeyCredential(acctName, acctKey)
+	}
+
+	pipelineOpts := azblob.PipelineOptions{
+		Telemetry: azblob.TelemetryOptions{
+			Value: "tegola-cache",
+		},
+	}
+
+	pipeline := azblob.NewPipeline(cred, pipelineOpts)
+
+	// container
+	uStr, err := config.String(ConfigKeyContainerUrl, nil)
+	if err != nil {
+		return nil, err
+	}
+	u, err := url.Parse(uStr)
+	if err != nil {
+		return nil, err
+	}
+
+	container := azblob.NewContainerURL(*u, pipeline)
+
+	azCache.Container = container
+
+	// in order to confirm we have the correct permissions on the container create a small file
+	// and test a PUT, GET and DELETE to the container
+	key := cache.Key{
+		MapName:   "tegola-test-map",
+		LayerName: "test-layer",
+		Z:         0,
+		X:         0,
+		Y:         0,
+	}
+
+	// seperate test for read only caches
+	if azCache.ReadOnly {
+		// read test file
+		_, _, err := azCache.Get(&key)
+		if err != nil {
+			e := cache.ErrGettingFromCache{
+				Err:       err,
+				CacheType: CacheType,
+			}
+
+			return nil, e
+		}
+	} else {
+		// write test file
+		err = azCache.Set(&key, []byte(testMsg))
+		if err != nil {
+			e := cache.ErrSettingToCache{
+				Err:       err,
+				CacheType: CacheType,
+			}
+
+			return nil, e
+		}
+
+		// read test file
+		byt, hit, err := azCache.Get(&key)
+		if err != nil {
+			e := cache.ErrGettingFromCache{
+				Err:       err,
+				CacheType: CacheType,
+			}
+
+			return nil, e
+		}
+		if !hit {
+			//return an error?
+			panic("no hit on writable cache")
+		}
+
+		// test response of writable cache
+		if string(byt) != testMsg {
+			e := cache.ErrGettingFromCache{
+				Err:       fmt.Errorf("incrrect test response %s != %s", string(byt), testMsg),
+				CacheType: CacheType,
+			}
+
+			return nil, e
+		}
+
+		err = azCache.Purge(&key)
+		if err != nil {
+			e := cache.ErrPurgingCache{
+				Err:       err,
+				CacheType: CacheType,
+			}
+
+			return nil, e
+		}
+	}
+
+	return &azCache, nil
+}
+
+type Cache struct {
+	Basepath  string
+	MaxZoom   uint
+	ReadOnly  bool
+	Container azblob.ContainerURL
+}
+
+func roundUp512(n int) int32 {
```

Comment:
needs test cases. currently does not handle the value 512 correctly (adds unnecessary padding)